### PR TITLE
fix(sync): Pause new downloads when Zebra reaches the lookahead limit

### LINKED ISSUE #5518: ci: logs longer than GitHub's limit do not allow to see the actual error in the console
## Motivation

Some error logs might be "hidden" if the logs are too long, needing extra steps for developer to find the actual error. And making it harder to _link_ the errors in issues/PRs

Example: https://github.com/ZcashFoundation/zebra/actions/runs/3332682537/jobs/5548148889

### Designs

We can approach this in two different ways:
- Confirm the actual limits on GitHub Actions, and just show the _end_ of the logs or the amount supported by GHA
- Make Zebra logs on certain tests less verbose

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

We're seeing a lot of AboveLookaheadLimit errors in full syncs. Instead of dropping those blocks, we should wait for them to verify.

Also we might need more time for a full sync.

### Designs

Add a watch channel to the downloader, which gets set when the downloader is a long way away from the state tip. Reset the channel when:
- a downloaded block is much closer to the tip, or
- there are no more blocks pending download or verify.

## Solution

Syncer fixes:
- Make the syncer pause new downloads when a downloaded block is 3000 blocks ahead of the state tip
    - If the block is 50,000 blocks ahead, still drop it anyway
- Make sure the default lookahead limit includes 2 full checkpoints (1000 blocks)
    - Add a new test config for this
- Allow 2000 extra blocks in the pre-download queue, or the state commit channel (was 4000 extra blocks)

Related fixes:
- Allow 36 hours for a full sync (was 32 hours)
- Only log lookahead issues once at info level for every pause, rather than for every block
    - Closes #5518

Related cleanups:
- Fix up a syncer test that didn't actually test what it said it did
- Add a `getblocktemplate-rpcs` alternative to the missing test config file message

## Review

Anyone can review this PR, it's a high priority because it fixes errors on `main`.

I'm doing a full sync locally to check that it works.

### Reviewer Checklist

  - [ ] Will the PR name make sense to users?
    - [ ] Does it need extra CHANGELOG info? (new features, breaking changes, large changes)
  - [ ] Are the PR labels correct?
  - [ ] Does the code do what the ticket and PR says?
  - [ ] How do you know it works? Does it have tests?

### CODE REVIEW

**Comment by teor2345** [comment]
@mergifyio update

**Comment by mpguerra** [comment]
Do we know why this has started happening now?

**Comment by teor2345** [comment]
> Do we know why this has started happening now?

Probably left over from the changes in #4937. We tried tweaking the lookahead limit a few times, but that doesn't seem to help.

**Comment by teor2345** [comment]
Syncing worked, but it was pausing for a block timeout every few checkpoints.
So I tweaked the syncer to start again when there were only a few blocks left, rather than waiting for zero blocks.

Apart from that, the logging and sync speed look really good.

**Comment by teor2345** [comment]
There are still some timeouts in the logs, but I'll deal with them in another PR.

File: zebrad/src/components/sync/downloads.rs [hunk 1]
```diff
@@ -42,14 +46,17 @@ type BoxError = Box<dyn std::error::Error + Send + Sync + 'static>;
...
 /// Since the syncer queue is limited to the `lookahead_limit`,
 /// the rest of the capacity is reserved for the other queues.
 /// There is no reserved capacity for the syncer queue:
 /// if the other queues stay full, the syncer will eventually time out and reset.
-pub const VERIFICATION_PIPELINE_SCALING_MULTIPLIER: usize = 5;
+pub const VERIFICATION_PIPELINE_SCALING_MULTIPLIER: usize = 2;
+
+/// The maximum height difference between Zebra's state tip and a downloaded block.
+/// Blocks higher than this will get dropped and return an error.
+pub const VERIFICATION_PIPELINE_DROP_LIMIT: i32 = 50_000;
```
Comment (arya2) [inline zebrad/src/components/sync/downloads.rs:59]: I've likely misunderstood how the lookahead limit works, but would it ever download a block that's 50k blocks above the tip?

File: zebrad/src/components/sync/downloads.rs [hunk 2]
```diff
@@ -259,15 +286,23 @@ where
         network: ZN,
         verifier: ZV,
         latest_chain_tip: ZSTip,
+        past_lookahead_limit_sender: watch::Sender<bool>,
```
Comment (arya2) [inline zebrad/src/components/sync/downloads.rs:289]: optional nit: an `AtomicBool` would be simpler (and insignificantly faster).

**Review by arya2 — COMMENTED** [review]
This is wonderful, thank you.

File: zebrad/src/components/sync.rs [hunk 3]
```diff
@@ -545,7 +555,14 @@ where
...
             self.update_metrics();
 
-            while self.downloads.in_flight() >= self.lookahead_limit(extra_hashes.len()) {
+            // Pause new downloads while the syncer or downloader are past their lookahead limits.
+            //
+            // To avoid a deadlock or long waits for blocks to expire, we ignore the download
+            // lookahead limit when there are only a small number of blocks waiting.
+            while self.downloads.in_flight() >= self.lookahead_limit(extra_hashes.len())
+                || (self.downloads.in_flight() >= self.lookahead_limit(extra_hashes.len()) / 2
+                    && self.past_lookahead_limit_receiver.cloned_watch_data())
```
Comment (dconnolly) [inline zebrad/src/components/sync.rs:564]: 👍

File: zebrad/src/components/sync/downloads.rs [hunk 4]
```diff
@@ -89,6 +96,14 @@ pub enum BlockDownloadVerifyError {
...
 
+    /// A downloaded block was a long way ahead of the state chain tip.
+    /// This error should be very rare during normal operation.
+    ///
+    /// We need to reset the syncer on this error, to allow the verifier and state to catch up,
+    /// or prevent it following a bad chain.
+    ///
+    /// If we don't reset the syncer on this error, it will continue downloading blocks from a bad
+    /// chain, or blocks far ahead of the current state tip.
     #[error("downloaded block was too far ahead of the chain tip: {height:?} {hash:?}")]
```
Comment (dconnolly) [inline zebrad/src/components/sync/downloads.rs:107]: 👍

File: zebrad/src/components/sync/downloads.rs [hunk 5]
```diff
@@ -388,28 +434,59 @@ where
...
-                    // it will continue downloading blocks from a bad chain,
-                    // (or blocks far ahead of the current state tip).
+                if block_height > lookahead_drop_height {
                     Err(BlockDownloadVerifyError::AboveLookaheadHeightLimit { height: block_height, hash })?;
-                } else if block_height < min_accepted_height {
+                } else if block_height > lookahead_pause_height {
+                    // This log can be very verbose, usually hundreds of blocks are dropped.
+                    // So we only log at info level for the first above-height block.
+                    if !past_lookahead_limit_receiver.cloned_watch_data() {
+                        info!(
```
Comment (dconnolly) [inline zebrad/src/components/sync/downloads.rs:443]: 👍

**Review by dconnolly — APPROVED** [review]

File: zebrad/src/components/sync/downloads.rs (hunk 1 above)
Comment (teor2345) [reply zebrad/src/components/sync/downloads.rs:59]: If we're 50,000 blocks ahead of the tip, we've either:
- had a bug in the syncer/state where it's not stopping even though it is a long way from the tip, or
- a peer gave us an incorrect block hash, or
- we misinterpreted some peer block hash message as a response to our `extend_tips()` query.

File: zebrad/src/components/sync/downloads.rs (hunk 2 above)
Comment (teor2345) [reply zebrad/src/components/sync/downloads.rs:289]: I tried this, but it needed an `Arc<AtomicBool>`, and making decisions around compare/exchange and atomic orderings.

It seemed simpler to do it this way.

**Comment by teor2345** [comment]
@mergifyio refresh

**Comment by teor2345** [comment]
@mergifyio refresh

**Pull request merged**